pandas/core/reshape/reshape.py

```diff
@@ -1053,6 +1053,168 @@ def get_empty_frame(data) -> DataFrame:
+        for col in data_to_decode.columns:
+            ps = [ps for ps in prefix_sep if ps in col][0]
+            prefix = col.split(ps)[0]
+            if prefix not in sep_for_prefix:
+                sep_for_prefix[prefix] = ps
+            if prefix not in variables_slice:
+                variables_slice[prefix] = [col]
+            else:
+                variables_slice[prefix].append(col)
+        prefix_sep = sep_for_prefix
+
+    # validate number of passed arguments
+    def check_len(item, name) -> None:
+        if not len(item) == len(variables_slice):
+            len_msg = (
+                f"Length of '{name}' ({len(item)}) did not match the "
+                "length of the columns being encoded "
+                f"({len(variables_slice)})."
+            )
+            raise ValueError(len_msg)
+
+    # obtain prefix to category mapping
+    variables: dict[str, str]
+    if isinstance(variables, dict):
+        check_len(variables, "variables")
+        variables = variables
+    elif is_list_like(variables):
+        check_len(variables, "variables")
+        variables = dict(zip(variables_slice, variables))
+    elif isinstance(variables, str):
+        variables = dict(
+            zip(
+                variables_slice,
+                (f"{variables}{i}" for i in range(len(variables_slice))),
+            )
+        )
+    else:
+        variables = dict(zip(variables_slice, variables_slice))
+
+    if dropped_first:
+        if isinstance(dropped_first, dict):
+            check_len(dropped_first, "dropped_first")
+        elif is_list_like(dropped_first):
+            check_len(dropped_first, "dropped_first")
+            dropped_first = dict(zip(variables_slice, dropped_first))
+        else:
+            dropped_first = dict(
+                zip(variables_slice, [dropped_first] * len(variables_slice))
+            )
+
+    cat_data = {var: [] for _, var in variables.items()}
+    for index, row in data.iterrows():
+        for prefix, prefix_slice in variables_slice.items():
+            slice_sum = row[prefix_slice].sum()
+            if slice_sum > 1:
+                raise ValueError(
+                    f"Dummy DataFrame contains multi-assignment(s) for prefix: "
+                    f"'{prefix}' in row {index}."
+                )
+            elif slice_sum == 0:
+                if dropped_first:
+                    category = dropped_first[prefix]
+                elif not dummy_na:
+                    category = np.nan
+                else:
+                    raise ValueError(
+                        f"Dummy DataFrame contains no assignment for prefix: "
+                        f"'{prefix}' in row {index}."
+                    )
+            else:
+                cat_index = row[prefix_slice].argmax()
+                category = prefix_slice[cat_index].split(prefix_sep[prefix])[1]
+                if dummy_na and category == "NaN":
+                    category = np.nan
```

Comment:
I'd say, let's not worry about adding `dummy_na` for now

Reply:
Without `dummy_na` and as there must be a way to handle unassigned values in the output `DataFrame` there are two options to deal with the `NaN` columns from [`get_dummies(data, dummy_na=True)`](https://pandas.pydata.org/docs/reference/api/pandas.get_dummies.html):

### Treat as category
- Treat the `prefix_NaN` column the same way as any other category is treated
   ```python
   >>> dummies
      C     col1_a  col1_b  col1_NaN  col2_a  col2_b  col2_c col2_NaN
   0  1       1       0        0        0       0       0       1
   1  2       0       1        0        1       0       0       0
   2  3       0       0        1        0       0       1       0

   >>> from_dummies(dummies, dummy_na=True)
      C      col1    col2 
   0  1       a      NaN
   1  2       b       a 
   2  3      NaN      c
   ```
- The variable is assigned a string `NaN` in the respective row.
- E.g. `col2` values can be element of `{a, b, c, NaN}`.
- Argument for this option is that matching a string to a part of a column name does not seem very robust and columns may contain a string `NaN` as a part of an actual catgory.

### Treat as category
- Ignore collumns with `prefix_NaN` 
   ```python
   >>> dummies
      C     col1_a  col1_b  col1_NaN  col2_a  col2_b  col2_c col2_NaN
   0  1       1       0        0        0       0       0       1
   1  2       0       1        0        1       0       0       0
   2  3       0       0        1        0       0       1       0

   >>> from_dummies(dummies, dummy_na=True)
      C      col1    col2 
   0  1       a     np.NaN
   1  2       b       a 
   2  3     np.NaN    c
   ```
   - With columns containing `prefix_NaN` being ignored some rows now have unassigned values which will then be treated as `np.nan`.
   - The justification to ignore these columns is drawn from the `prefix_NaN` column generation in `get_dummies(data, dummy_na=True)` which makes this column name some sort of special case.
   - The result would be similar to the verion with `dummy_na` but the the `np.nan` value would be drawn from an unassigned row which seems to be a cleaner implementaiton.

Reply:
For now I'd suggest to treat it as any other column and to remove `dummy_na`, the extra complexity can be added later if desired

Reply:
Sounds good